pkg/util/hostname/file/hostname.go

```diff
@@ -0,0 +1,43 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package file
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/util/hostname/validate"
+)
+
+// HostnameProvider parses a file from the 'filename' option and returns the content
+// as the hostname
+func HostnameProvider(ctx context.Context, options map[string]interface{}) (string, error) {
```

Comment:
Can you elaborate a little bit on your vision for `options` as part of the Provider interface? I don't really mind it, as I do see some use for it to avoid import cycles, but I'd like to understand what you have in mind. Right now it feels like it's quite specific to this use case with `file.HostnameProvider`, and could maybe be addressed differently, so I want to make sure it makes sense.

Reply:
@truthbk Sure! So There are some spots ([example](https://github.com/DataDog/datadog-agent/blob/bd717d5fc0ed2730561c9faab7ddaef66266d655/pkg/util/hostname.go#L276), [example2](https://github.com/DataDog/datadog-agent/blob/bd717d5fc0ed2730561c9faab7ddaef66266d655/pkg/util/hostname.go#L242-L243), [example3](https://github.com/DataDog/datadog-agent/blob/bd717d5fc0ed2730561c9faab7ddaef66266d655/pkg/util/hostname.go#L223-L239)) in this logic where we seem to leak specific abstractions that should be isolated in a provider back into this hostname resolution logic itself. If we allowed options to be passed into the providers, I think it would allow us to encapsulate all of the special logic paths into "providers" that we just loop over and remove this dense/complex glue logic. While the file provider seems to be the first one to use this parameter directly, having `options` would reduce shoe-horning of future code to span between the providers and `hostname.go` just because there was no way to pass in any options to the provider itself.

Reply:
My only concern with this is that it's hard for the caller to know what arguments to give it without looking at the implementaion. Though I can't think of a much better way to do this without adding a lot of boilerplate code. Just wanted to mention it since I generally think strict interfaces are a good thing.

Reply:
> I generally think strict interfaces are a good thing

Yeah I would agree with you on that completely but given the diversity of our hostname providers, I don't know if we could have a common set of options as a struct that would make sense for all providers without some flexible way to add provider-specific params.